switcheroo/data_stores.py

```diff
@@ -0,0 +1,161 @@
+"Data stores that specifies where a Server stores its keys"
+from abc import ABC, abstractmethod
+from pathlib import Path
+from tempfile import TemporaryDirectory
+import boto3
+from switcheroo import paths
+from switcheroo.custom_keygen import KeyGen, KeyMetadata
+
+
+class DataStore(ABC):
+    "A server uses a DataStore to get public keys from somewhere."
+
+    def __init__(self, ssh_home: Path | None = None, temp: bool = False):
+        # If temp is true, will reset the home_dir to a temp file upon usage as a context manager
+        self.temp = temp
+        # ssh_home is only relevant if not being used as a context manager
+        self._dir: Path = paths.local_ssh_home() if ssh_home is None else ssh_home
+        # To be used later if we decide to use this as a context manager, and temp is selected
+        self._temp_dir: TemporaryDirectory[str] | None = None
+
+    @property
+    def home_dir(self) -> Path:
+        """The folder where local files are stored"""
+        if self.temp and self._temp_dir is not None:
+            return Path(self._temp_dir.name)
+        return self._dir
+
+    @abstractmethod
+    def get_sshd_config_line(self):
+        "Return the config line that the server will add to the sshd_config"
+
+    @abstractmethod
+    def publish(self, host: str, user: str) -> str:
+        """Publish a new public key for the given host and user"""
+
+    @abstractmethod
+    def publish_with_metadata(
+        self, host: str, user: str, metadata: KeyMetadata | None
+    ) -> tuple[str, KeyMetadata]:
+        """Publish a new public key for the given host and user with metadata"""
+
+    @abstractmethod
+    def retrieve(self, host: str, user: str) -> str:
+        """Retrieve the public key for the given host and user"""
+
+    def __enter__(self):
+        if self.temp:
+            self._temp_dir: TemporaryDirectory[str] | None = TemporaryDirectory[
+                str
+            ](  # pylint: disable=consider-using-with
+                prefix="ssh-keys-", dir=self._dir
+            )
+
+    def __exit__(self, exc_t, exc_v, exc_tb):
+        if self.temp:
+            assert self._temp_dir is not None
+            self._temp_dir.__exit__(None, None, None)
+
+
+class S3DataStore(DataStore):
+    "Store the public keys in an S3 bucket"
+
+    def __init__(
+        self, _s3_bucket_name: str, ssh_home: Path | None = None, temp: bool = False
+    ):
+        super().__init__(ssh_home=ssh_home, temp=temp)
+        self._s3_bucket_name = _s3_bucket_name
+
+    @property
+    def s3_bucket_name(self):
+        "The name of the bucket in which the keys are stored"
+        return self._s3_bucket_name
+
+    def get_sshd_config_line(self) -> str:
```

Comment:
this doesn't seem to belong here it has nothing to do with the data store

Reply:
Are you referring to the sshd_config_line method? If so I see, we can extract that.

Reply:
However, would it then just be a check in the Server class? ie the server checks which data store is being used and generates the command line accordingly? If so, then we lose abstraction